### PR TITLE
[Merged by Bors] - feat(CharP): `Int.cast` is injective on `Ico 0 p`

### PR DESCRIPTION
**Pull request by YaelDillies**
and generalise `Fin.charP`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/CharP/Defs.lean [hunk 1]
```diff
@@ -91,6 +93,13 @@ end AddMonoidWithOne
 section AddGroupWithOne
 variable [AddGroupWithOne R] (p : ℕ) [CharP R p] {a b : ℤ}
 
+lemma intCast_injOn_Ico [IsRightCancelAdd R] : InjOn (Int.cast : ℤ → R) (Ico 0 p) := by
```
Comment (urkud) [inline Mathlib/Algebra/CharP/Defs.lean:96]: Should we have this for any `Ico a (a + p)`?

File: Mathlib/Algebra/CharP/Defs.lean [hunk 2]
```diff
@@ -411,7 +420,7 @@ end CharZero
 
 namespace Fin
 
-instance charP (n : ℕ) : CharP (Fin (n + 1)) (n + 1) where
-    cast_eq_zero_iff' := by simp [Fin.ext_iff, Nat.dvd_iff_mod_eq_zero]
+instance charP (n : ℕ) [NeZero n] : CharP (Fin n) n where
```
Comment (urkud) [inline Mathlib/Algebra/CharP/Defs.lean:423]:
```suggestion
instance instCharP (n : ℕ) [NeZero n] : CharP (Fin n) n where
```

File: Mathlib/Algebra/CharP/Defs.lean [hunk 3]
```diff
@@ -411,7 +420,7 @@ end CharZero
...
+  cast_eq_zero_iff' := by simp [Fin.ext_iff, Nat.dvd_iff_mod_eq_zero]
```
Comment (urkud) [inline Mathlib/Algebra/CharP/Defs.lean:424]: Is `Fin.natCast_eq_zero` available here?

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**